src/Buckle/Buckle/CodeAnalysis/Parser/Blender.Reader.cs

```diff
@@ -0,0 +1,133 @@
+                    return ReadNewToken();
+
+                if (_changeDelta < 0) {
+                    SkipOldToken();
+                } else if (_changeDelta > 0) {
+                    return ReadNewToken();
+                } else {
+                    if (TryTakeOldNodeOrToken(asToken, out var blendedNode))
+                        return blendedNode;
+
+                    if (!_oldTreeCursor.currentNode.kind.IsToken())
+                        _oldTreeCursor = _oldTreeCursor.MoveToFirstChild();
+                    else
+                        SkipOldToken();
+                }
+            }
+        }
+
+        private void SkipOldToken() {
+            _oldTreeCursor = _oldTreeCursor.MoveToFirstToken();
+            var node = _oldTreeCursor.currentNode;
+
+            _changeDelta += node.fullSpan.length;
+            _oldTreeCursor = _oldTreeCursor.MoveToNextSibling();
+
+            SkipPastChanges();
+        }
+
+        private void SkipPastChanges() {
+            var oldPosition = _oldTreeCursor.currentNode.fullSpan.start;
+
+            while (!_changes.IsEmpty && oldPosition >= _changes.Peek().span.end) {
+                var change = _changes.Peek();
+                _changes = _changes.Pop();
+                _changeDelta += change.newLength - change.span.length;
+            }
+        }
+
+        private BlendedNode ReadNewToken() {
+            var token = LexNewToken();
+            var width = token.fullSpan.length;
+            _newPosition += width;
+            _changeDelta -= width;
+
+            SkipPastChanges();
+
+            return CreateBlendedNode(null, token);
+        }
+
+        private SyntaxToken LexNewToken() {
+            if (_lexer.position != _newPosition)
+                _lexer.Move(_newPosition);
+
+            var token = _lexer.LexNext();
+            return token;
+        }
+
+        private bool TryTakeOldNodeOrToken(bool asToken, out BlendedNode blendedNode) {
+            if (asToken)
+                _oldTreeCursor = _oldTreeCursor.MoveToFirstToken();
+
+            var node = _oldTreeCursor.currentNode;
+
+            if (!CanReuse(node)) {
+                blendedNode = null;
+                return false;
+            }
+
+            var oldPosition = _newPosition;
+            _newPosition += node.fullSpan.length;
+            _oldTreeCursor = _oldTreeCursor.MoveToNextSibling();
+
+            blendedNode = CreateBlendedNode(node.kind.IsToken() ? null : node, node as SyntaxToken);
+            return true;
+        }
+
+        private bool CanReuse(SyntaxNode node) {
+            if (node.fullSpan.length == 0)
+                return false;
+
+            if (IntersectsNextChange(node))
+                return false;
+
+            if (node.kind.IsToken() && (node as SyntaxToken).isFabricated)
+                return false;
+
+            if (node.GetLastToken().isFabricated)
+                return false;
+
+            return true;
```

Comment:
Could maybe attach relevant diagnostics to each node in the tree and then use those diagnostics to make sure the node can be reused without shadowing some errors that the parser or lexer would create by using it.